sql/catalog: avoid fetching descriptors when fetching comments

fqazi opened the pull request:
Previously, we refactored the code to fetch comments, descriptors, and zone config together in all cases. A side effect of this change was that the crdb_internal.kv_system_comments table was substantially slower for larger tables leading to big regressions. To address this, this patch adds a method of only fetching comments within the collections.

Informs: #102851
Fixes: #102613

Release note (bug fix): Optimize over-head of
pg_catalog.pg_description and pg_catalog.pg_shdescription, which can lead to performance regression relative to 22.2

Commit:
Merge #102593

102593: c2c: skip and deflake a few unit tests r=benbardin a=msbutler

c2c: skip a few unit tests
Informs #102523
Informs #102286
Informs #86206

Release note: none

c2c: close pgconn with correct context in test infra
When a user opens a pgx connection, they pass it a context that must be used to
close the connection, else pgx's contextWatcher will run indefinetly on a
seperate goroutine.  Previously, the `startReplication()` test helper was
initializing its own context to open pgx connection, and did not pass this
context back to the caller, and consequently, the user would close the
connection with the incorrect context, leading to a leaky goroutine.

This patch allows the user to pass a context in the startReplication(),
preventing the user from closing the connection with the wrong goroutine.

Fixes https://github.com/cockroachdb/cockroach/issues/102523

Release note: None

Co-authored-by: Michael Butler <butler@cockroachlabs.com>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/102991)
<!-- Reviewable:end -->

fqazi committed:
sql/catalog: avoid fetching descriptors when fetching comments

Previously, we refactored the code to fetch comments, descriptors,
and zone config together in all cases. A side effect of this change
was that the crdb_internal.kv_system_comments table was substantially
slower for larger tables leading to big regressions. To address this,
this patch adds a method of only fetching comments within the
collections.

Informs: #102851
Fixes: #102613

Release note (bug fix): Optimize over-head of
pg_catalog.pg_description and pg_catalog.pg_shdescription, which
can lead to performance regression relative to 22.2

chengxiong-ruan commented on pkg/sql/catalog/internal/catkv/catalog_reader_cached.go:
```diff
@@ -149,6 +150,26 @@ func (c *cachedCatalogReader) Reset(ctx context.Context) {
+	if err != nil {
+		return nstree.Catalog{}, err
+	}
+	/*if err := c.ensure(ctx, read); err != nil {
```
nit: remove these

chengxiong-ruan reviewed (approved):
lgtm

fqazi commented:
@chengxiong-ruan TFTR!

rafiss reviewed (changes requested):
nice fix. can you add a test for this in rttanalysis?

rafiss reviewed (commented):
we should be able to use the query that dikshant added in the ticket. it came from an ORM, so it can be in the ORMQueries benchmark test.
```
SELECT
  current_database(), n.nspname, relname, d.description
FROM
  pg_description AS d
  INNER JOIN pg_class AS c ON d.objoid = c.oid
  INNER JOIN pg_namespace AS n ON n.oid = c.relnamespace
WHERE
  d.objsubid = 0
  AND n.nspname
    NOT IN (
        'gp_toolkit':::STRING:::NAME,
        'information_schema':::STRING:::NAME,
        'pgagent':::STRING:::NAME
      )
  AND n.nspname NOT LIKE 'pg_%';
```

fqazi commented:
> we should be able to use the query that dikshant added in the ticket. it came from an ORM, so it can be in the ORMQueries benchmark test.
> 
> ```
> SELECT
>   current_database(), n.nspname, relname, d.description
> FROM
>   pg_description AS d
>   INNER JOIN pg_class AS c ON d.objoid = c.oid
>   INNER JOIN pg_namespace AS n ON n.oid = c.relnamespace
> WHERE
>   d.objsubid = 0
>   AND n.nspname
>     NOT IN (
>         'gp_toolkit':::STRING:::NAME,
>         'information_schema':::STRING:::NAME,
>         'pgagent':::STRING:::NAME
>       )
>   AND n.nspname NOT LIKE 'pg_%';
> ```

@rafiss  weirdly rttanalysis chokes on this query. This one is bad; it takes 11 seconds on a  cockroach demo. I used the one from Django:

```
SELECT
                c.relname,
                CASE
                    WHEN c.relispartition THEN 'p'
                    WHEN c.relkind IN ('m', 'v') THEN 'v'
                    ELSE 't'
                END,
                obj_description(c.oid, 'pg_class')
            FROM pg_catalog.pg_class c
            LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
            WHERE c.relkind IN ('f', 'm', 'p', 'r', 'v')
                AND n.nspname NOT IN ('pg_catalog', 'pg_toast')
                AND pg_catalog.pg_table_is_visible(c.oid);
```

fqazi commented:
Let me dig into the regression with the other one

fqazi committed:
pg_catalog: add virtual indexes and optimize existing ones

Previously, the pg_description table was missing a virtual
index on objoid, with the addition of descriptions from all
builtins this table has much more data by default. As a result
any query joining with this table is slow. To address this,
this patch will first add a virtual index on the index
description table, and update existing incomplete indexes
on OID's to skip over builtins that refer to builtin functions.

Release note (performance improvement): improve performance
when joining with the pg_description table
